sybil/example.py

```diff
@@ -51,15 +51,15 @@ def __init__(
         self.parsed: Any = region.parsed
         #: The :attr:`~sybil.Document.namespace` of the document from
         #: which this example came.
-        self.namespace: dict = namespace
+        self.namespace: Dict[str, Any] = namespace
 
     def __repr__(self) -> str:
         return '<Example path={} line={} column={} using {!r}>'.format(
             self.path, self.line, self.column, self.region.evaluator
         )
 
     def evaluate(self) -> None:
-        evaluator = self.document.evaluator or self.region.evaluator
+        evaluator = self.document.evaluator if self.document.evaluator is not None else self.region.evaluator
```

Comment:
I think `mypy` doesn't like checking truthiness of a function:

```
sybil/example.py:62: error: Function "evaluator" could always be true in boolean context  [truthy-function]
```

Reply:
`Function "evaluator" could always be true in boolean context` is a pretty crappy message :-(
Do you have a link to the mypy docs on what it's sad about here?

The change is needlessly wordy and makes me sad about appeasing mypy's inadequacies :-(

Reply:
The error was related to https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-function-isn-t-used-in-boolean-context-truthy-function, because `self.document.evaluator` was typed as an `Evaluator`.
I have changed it to be correctly typed as an `Optional[Evaluator]`, and therefore been able to revert the change which made you sad.

That said, we will hit into related problems with `self.document.evaluator` later, as `mypy` does not allow assigning to a method:

```python
from typing import Callable, Optional

class Example:
    def __init__(self) -> None:
        # mypy gives "error: Cannot assign to a method  [method-assign]" for this.
        self.foo: Optional[Callable[[], None]] = None

    def foo(self) -> None:
        pass
```

I think that in order to satisfy `mypy`, we will either have to add `type: ignore` s (which I try to avoid, where possible), or change `document` to be more complex in some ways (likely having another variable to track whether the `evaluator` should be used).

Reply:
That "can't assign to a method" feels like a straight up in mypy, no? 
I'm guessing this problem has come up with other folks attempting to add type hints to existing code of this type?

Reply:
> That "can't assign to a method" feels like a straight up in mypy, no?

I'm not sure what that means

Reply:
Sorry, braino in sentence on my part: can you have a look and see if there's a mypy bug logged for this already and if not, log one? I don't want a `type: ignore` and I certainly don't want to change `document`. It's perfectly legit to assign to a method in Python (Sybil has 100% test coverage to prove it ;-)).

Reply:
There are many discussions on this - with many people wanting `mypy` to support assigning to a method. See https://github.com/python/mypy/issues/708 for an example.

I do not get the impression that very soon there will me a `mypy` change to accommodate. You can even see an issue raised by me in 2018 when I wanted to assign to a method precisely as described in the Python documentation: https://github.com/python/mypy/issues/5062#issue-323652082.

In a follow-up PR, I think we will have to decide either to change `document`, to add a `type: ignore`, or to ignore all `method-assign` errors.

Reply:
Can this check be disabled in mypy's config file?

Reply:
> Can this check be disabled in mypy's config file?

Yes, that's what I meant when I said: "or to ignore all `method-assign` errors.".

Any specific `mypy` error code can be ignored in `mypy`'s config file.

Reply:
Okay, let's do that... Might be worth getting a branch going that has mypy turned on and rebase that onto master as more PRs land. 

(ie: just the CI, so we know where we're at and what config we'll be using...)

That said, how hard would it be to use pyright instead of mypy? I think @hynek has had some success with this?

Reply:
@cjw296 I'm happy to do those tasks. I think they're best as follow-ups, as this change doesn't yet hit the problem:

* Add `mypy` to CI, with `pypyoject.toml` level ignores for files which are not yet compliant
* Address the `method-assign` error. From this conversation, most likely by globally ignoring that error type. I will look into that further in the follow-up, in case there are complications I have not considered here.
* Look into adding `pyright` (potentially as a replacement for `mypy`). I will say that in my own projects, I use both `pyright` and `mypy`, and I have found roughly equal numbers of annoyances. I like having both as, despite annoyances, they each have caught a few bugs that the other has not.